File: java/P15.java
// P15 Replicate the elements of a list a given number of times.

import core.List;

import static core.List.*;

class P15 {
    // Recursion, one-pass concatenation
    <T> List<T> f1(int n, List<T> list) {
        return list.isEmpty()
                ? List.<T>nil()
                : prepend(n, list.head(), f1(n, list.tail()));
    }

    <T> List<T> prepend(int n, T x, List<T> list) {
        return n == 0 ? list : cons(x, prepend(n - 1, x, list));
    }

    // Imperative
    <T> List<T> f2(int n, List<T> list) {
        List<T> acc = nil();
        for (T each : list)
            for (int i = 0; i < n; i++)
                acc = cons(each, acc);
        return acc.reverse();
    }
}

File: clojure/P15.clj
; P15 Replicate the elements of a list a given number of times.

; Replicate then mapcat
(defn f1 [n l]
  (mapcat #(repeat n %) l))

; Without using "repeat", one-pass concatenation
(defn f2 [n l]
  (defn prepend [n x xs]
    (if (= n 0) xs (cons x (prepend (dec n) x xs))))
  (reduce #(prepend n %2 %1) '() (reverse l)))